# **Tarea 1 - CC6205 Natural Language Processing 📚**

**Integrantes: Vicente Ardiles S. - Rodrigo I. Oportot**

**Fecha límite de entrega 📆:** Lunes 12 de Abril.

**Tiempo estimado de dedicación: 6hrs**




` ` 


Bienvenid@s a la primera tarea en el curso de Natural Language Processing (NLP). Esta tarea tiene como objetivo evaluar los contenidos teóricos de las primeras semanas de clases, enfocado principalmente en ***Information Retrieval (IR)*** y ***Vector Space Models***. Si aún no han visto las clases, se recomienda visitar los links de las referencias.

La tarea consta de una parte teórica que busca evaluar conceptos vistos en clases. Seguido por una parte práctica con el fín de introducirlos a la programación en Python enfocada en NLP. 

` ` 




**Instrucciones:**
- La tarea se realiza en grupos de **máximo** 2 personas. Puede ser invidivual pero no es recomendable.
- La entrega es a través de u-cursos a más tardar el día estipulado arriba. No se aceptan atrasos.
- El formato de entrega es este mismo Jupyter Notebook.
- Al momento de la revisión tu código será ejecutado. Por favor verifica que tu entrega no tenga errores de compilación. 
- Está **PROHIBIDO** usar cualquier librería que implemente los algoritmos pedidos (Spacy, scikit, etc). Sólo se podrán utilizar las librerías importadas al inicio de la sección de práctica.
- En el horario de auxiliar pueden realizar consultas acerca de la tarea a través del canal de Discord del curso. 



Ahora sí, empecemos! 😄🚀

` ` 

**Referencias:**


Slides:
    
- [Introducción al curso](https://github.com/dccuchile/CC6205/blob/master/slides/NLP-introduction.pdf)
- [Vector Space Model / Information Retrieval](https://github.com/dccuchile/CC6205/blob/master/slides/NLP-IR.pdf)    

Videos: 

- [CC6205 - Procesamiento de Lenguaje Natural: CC6205 - Procesamiento de Lenguaje Natural: Introducción parte I](https://www.youtube.com/watch?v=HEKTNOttGvU)
- [CC6205 - Procesamiento de Lenguaje Natural: CC6205 - Procesamiento de Lenguaje Natural: Introducción parte II](https://www.youtube.com/watch?v=P8cwnI-f-Kg)
- [CC6205 - Procesamiento de Lenguaje Natural: Vector Space Model and Information Retrieval parte 1](https://youtu.be/FXIVClF370w)
- [CC6205 - Procesamiento de Lenguaje Natural: Vector Space Model and Information Retrieval parte 2](https://youtu.be/f8nG1EMmPZk)

` ` 



## **1 - Preguntas teóricas 📕 (2 puntos).** ##









Las siguientes celdas contienen preguntas acerca del contenido visto en clases y en el material del curso.  Contestar cada pregunta en su celda correspondiente y **no extenderse más de 5 lineas** . 🙏
` ` 

**Pregunta 1 (0.5 puntos): ¿Por qué el análisis del lenguaje humano es una tarea compleja? Mencione dos razones.**



` ` 

    Respuesta: 
    Una razón es su ambigüedad, uno puede agregar o quitar una palabra de una oración y cambiar totalmente el contexto y sentido de ella, sumado a que el lenguaje evoluciona con el tiempo, pudiendo no tener el mismo significado según la época. Por un lado más técnico, está la morfología de las palabras, no se pudiendo inferir una relación semántica entre palabras por las letras que la componen (casa y caza, por ejemplo).


` ` 

**Pregunta 2 (0.5 puntos): ¿Cuál es la diferencia entre Natural Language Processing (NLP) y Computational Linguistics?**

` ` 

    Respuesta:
    PLN se encarga de la metodología y diseños para resolver problemas (tasks) concretos que involucren lenguaje natural, con algoritmos y representaciones que lo trabajen. Computational Linguistics es el estudio de los procesos computacionales asociados al entendimiento del lenguaje natural, tratando de entender el lenguaje desde un punto de vista científico/lingüista/computacional.
    
` ` 

**Pregunta 3 (0.5 puntos): ¿Qué es el Machine Learning supersivado? ¿Por qué es importante para el campo de NLP?**

` ` 

    Respuesta:
    Son técnicas de ML donde el aprendizaje de los algoritmos que resolverán las tasks son estudiados e intervenidas a medida que se desarrollan, construyendo las reglas del lenguaje estudiado a través de un conjunto de datos ya conocidos (entrenamiento), etiquetados previamente. Destaca en NLP dada la dificultad de formalizar el lenguaje natural, siendo una herramienta muy útil para abordar tal problema.

` ` 

**Pregunta 4 (0.5 puntos): ¿Cuáles son las diferencias entre usar Deep learning y Machine Learning clásico (empirismo) para un problema de NLP? Ejemplifique con alguna task.**

` ` 

    Respuesta:
    El empirismo se caracterizaba por explotar manualmente el corpus a estudiar, luego usando modelos de ML supervisado y conceptos estadísticos. DL reemplaza lo anterior con algoritmos que aprenden sobre la marcha las reglas del lenguaje procesado. Ejemplo: NER con empirismo trabaja los datos crudos extrayendo atributos para procesarlos con ML supervisado, en cambio, en DL los algoritmos hacen la mayor parte del trabajo sin ese pre-procesamiento.
    

` ` 

## **2 - Preguntas prácticas 💻 (4 puntos).** ##

Esta segunda sección incluye ejercicios de programación 🤙. Leer atentamente las instrucciones entregadas a continuación para facilitar el proceso de revisión de sus trabajos.


` ` 

**Instrucciones:**



- Escribe tu código entre las lineas de comentarios **### Aquí inicia tu código ###** y **### Aquí termina tu código ###**.
- Cuando el ejercicio incluya un bloque llamado ***Test***, comprueba que el resultado de la ejecución coincida con el resultado esperado.
- Recuerde siempre mantener buenas prácticas de código.
- Está permitido sólo utilizar las librerías importadas antes del Ejercicio 1.
- **Recordar** que: *Documento = Oración. Dataset = Corpus. Vocabulario = Typos*.




**Ejemplo:** Implemente una función **`hello_world()`** que imprima en pantalla `"Hello World"`. 

In [1]:
def hello_world():
  ### Aquí inicia tu código ###
  print("Hello World")
  ### Aquí termina tu código ###

***Test:***

In [2]:
hello_world()

Hello World


***Resultado esperado***: 
<table>
    <tr> 
        <td> Hello World </td> 
    </tr>
</table> 

``
``




Estas son las librerías permitidas. Si quieren utilizar alguna librería adicional, pueden realizar la consulta a través de Discord. 

In [3]:
import math
import codecs
import re
import numpy as np
import pandas as pd

` ` 

**Ejercicio 1 - *Tokenización* (0.5 puntos).** 
` `  
` ` 

En el primer ejercicio veremos la dificultad 😨 de tokenizar textos no estructurados, destacando la importancia de tener librerías que realicen este trabajo. 

El archivo adjunto al enunciado de la tarea contiene una interconsulta médica real en Chile, estos textos son frecuentemente utilizados para obtener información epidemiológica y realizar estudios estadísticos a partir de ellos. Es por esto que poder analizar el texto de manera correcta es fundamental. 


Ejecute el código a continuación para cargar el ejemplo. Recuerde realizar la modificación al directorio en caso que el archivo no se encuentre en el mismo directorio del Jupyter Notebook.

In [4]:
text = codecs.open('/content/sample_data/interconsulta.txt', 'r', 'UTF-8').read()
print(text)

Interconsulta n°2 del mes.-05/10/2019
- ENFERMEDAD RENAL  CRÓNICA ETAPA 3. 
- Fundamento Clínico APS: PACIENTE  DE 61 AÑOS, 	CON ANTECEDENTES DE HRTA DE LARGA DATA.ACUDE A CONTROL PSCV CON EXAMNES, DONDE DESTACA CREATININA DE 1,43,
 CON UNA VFG CALCULADA DE 53 ML/MIN/1.73M2. CON UNA CREATININA HACE MAS DE 3 MESES DE 1.34.
                   
Enfermedad renal cronica Etapa 3(SOLICITO EVALUACION DEL CASO).



Implementen una función **`get_tokens()`** que reciba un texto y entregue una lista con sus tokens. Son libres de elegir la forma de tokenizar mientras no utilicen librerías con tokenizadores ya implementados. Pueden utilizar la librería **re** importada para trabajar símbolos.


Ejemplo de uso:

`get_tokens('Este es un ejemplo de prueba.')` 

Nos entregaría:

`['Este', 'es', 'un', 'ejemplo', 'de', 'prueba', '.']`

In [5]:
def get_tokens(text):
  ### Inicio del código ###
  
  output = []
  regex = []

  """
  Con expresiones regulares, trataremos de tokenizar:
  - los numeros con parte decimal
  - enteros sin separar sus digitos
  - palabras
  - simbolos como . , ? - entre otros
  """

  ## decimal o entero
  numbers = "\d+\.?,?\d+"
  regex.append(numbers)

  word = "[0-9]*[A-Za-zñáéíóúÁÉÍÓÚ]+[0-9]*"
  regex.append(word)
  
  symbol = "[.,\?\-_\(\)°/]+"
  regex.append(symbol)

  ## usar re.findall(pattern, string, flags=0)
  ## primero, por cada regex
  for item1 in regex:
    found_words = re.findall(item1, text, flags=0)

    ## se colocan las expresiones que hicieron match en el output
    for item2 in found_words:
      output.append(item2)
  
  return output
  ### Fín del código ###

In [6]:
tokens = get_tokens(text)
tokens

['05',
 '10',
 '2019',
 '61',
 '1,43',
 '53',
 '1.73',
 '1.34',
 'Interconsulta',
 'n',
 'del',
 'mes',
 'ENFERMEDAD',
 'RENAL',
 'CRÓNICA',
 'ETAPA',
 'Fundamento',
 'Clínico',
 'APS',
 'PACIENTE',
 'DE',
 'A',
 'OS',
 'CON',
 'ANTECEDENTES',
 'DE',
 'HRTA',
 'DE',
 'LARGA',
 'DATA',
 'ACUDE',
 'A',
 'CONTROL',
 'PSCV',
 'CON',
 'EXAMNES',
 'DONDE',
 'DESTACA',
 'CREATININA',
 'DE',
 'CON',
 'UNA',
 'VFG',
 'CALCULADA',
 'DE',
 'ML',
 'MIN',
 '73M2',
 'CON',
 'UNA',
 'CREATININA',
 'HACE',
 'MAS',
 'DE',
 'MESES',
 'DE',
 'Enfermedad',
 'renal',
 'cronica',
 'Etapa',
 'SOLICITO',
 'EVALUACION',
 'DEL',
 'CASO',
 '°',
 '.-',
 '/',
 '/',
 '-',
 '.',
 '-',
 ',',
 '.',
 ',',
 ',',
 ',',
 '/',
 '/',
 '.',
 '.',
 '.',
 '.',
 '(',
 ').']

**Describa cuáles fueron sus supuestos para realizar la tokenización y compare sus tokens con los entregados por la librería nltk en el bloque de código de más abajo.**

    Describa aquí
    Se supuso que pueden haber palabras con números al comienzo o al final de
    ellas, como en el regex word del cuerpo de la función. También que las
    fracciones siempre se escribirán como números con puntos o comas entre la
    parte entera y decimal. Los símbolos especiales asociados son los escritos
    en el regex symbol.

    Al comparar los tokens con los de nltk, se puede notar que el algoritmo de
    la libreria entrega los tokens en el orden en que aparecen en el texto,
    contrario a lo que hace la función get_tokens() definida más arriba. Debe
    tener más sentido sintáctico, pero para efectos de una lista desordenada,
    es muy parecido. 

In [7]:
from nltk.tokenize import wordpunct_tokenize # Nltk no puede ser utilizada para los siguientes ejercicio, es sólo un ejemplo
nltk_tokens = wordpunct_tokenize(text)
nltk_tokens

['Interconsulta',
 'n',
 '°',
 '2',
 'del',
 'mes',
 '.-',
 '05',
 '/',
 '10',
 '/',
 '2019',
 '-',
 'ENFERMEDAD',
 'RENAL',
 'CRÓNICA',
 'ETAPA',
 '3',
 '.',
 '-',
 'Fundamento',
 'Clínico',
 'APS',
 ':',
 'PACIENTE',
 'DE',
 '61',
 'AÑOS',
 ',',
 'CON',
 'ANTECEDENTES',
 'DE',
 'HRTA',
 'DE',
 'LARGA',
 'DATA',
 '.',
 'ACUDE',
 'A',
 'CONTROL',
 'PSCV',
 'CON',
 'EXAMNES',
 ',',
 'DONDE',
 'DESTACA',
 'CREATININA',
 'DE',
 '1',
 ',',
 '43',
 ',',
 'CON',
 'UNA',
 'VFG',
 'CALCULADA',
 'DE',
 '53',
 'ML',
 '/',
 'MIN',
 '/',
 '1',
 '.',
 '73M2',
 '.',
 'CON',
 'UNA',
 'CREATININA',
 'HACE',
 'MAS',
 'DE',
 '3',
 'MESES',
 'DE',
 '1',
 '.',
 '34',
 '.',
 'Enfermedad',
 'renal',
 'cronica',
 'Etapa',
 '3',
 '(',
 'SOLICITO',
 'EVALUACION',
 'DEL',
 'CASO',
 ').']

` `  
` ` 

**Ejercicio 2 - *Stopwords y Stemming* (1 punto).** 
` `  
` ` 

Considere el siguiente corpus:


In [8]:
dataset = ["I like human languages", "I like programming languages", "Spanish is my favorite language"]

Diseñe una función **`get_vocab()`** que extraiga los typos de este corpus solamente tokenizando. Puede utilizar la función creada por ustedes en el  Ejercicio 1.

In [9]:
def get_vocab(dataset):
  ### Aquí inicia tu código ###
  """
  Primero vamos a tokenizar cada documento 
  y luego añadir cada token a la lista output
  """
  output = []

  for var in dataset:
    tokenized = get_tokens(var)
    
    for token in tokenized:
      ## revisamos que no se repitan
      if(token not in output):
        output.append(token)

  return output

  ### Aquí termina tu código ###

***Test:***

In [10]:
vocab = get_vocab(dataset)
vocab

['I',
 'like',
 'human',
 'languages',
 'programming',
 'Spanish',
 'is',
 'my',
 'favorite',
 'language']

``
``

***Resultado esperado***: 
<table>
    <tr> 
        <td>['favorite',
 'Spanish',
 'language',
 'I',
 'like',
 'programming',
 'languages',
 'my',
 'human',
 'is'] </td> 
    </tr>
</table> 

``
``

Ahora diseñe reglas que usted estime convenientes tanto de **Stemming** como de **Stopwords**. Implemente una función que reciba una lista con los elementos del vocabulario, le aplique sus reglas y devuelva el vocabulario actualizado. 

    Explique sus reglas aquí

    Primero, las stopwords se eliminan, comparándolas en base a una lista de stopwords genérica en inglés.

    Segundo, se revisarán las terminaciones de los strings (los sufijos de las 
    palabras), por ejemplo, en español la palabra "perros" debería ser modificada
    a "perro". Las palabras cuyos finales sean cortados pueden repetirse, por ejemplo "perros" y "perritos"
    son cortadas a "perro", por lo que solamente se añade una vez al resultado final. 

    Se busca implementar una versión más sencilla del algoritmo de Porter, con el fin de no crear
    más funciones auxiliares, quedando la simplificación del proceso dentro de pre_processing().
    No se modelará cada palabra en base a la cadena \[C] (VM)^n \[V], sino que solamente se revisará 
    el fin de cada string junto con las reglas, para luego modificarlo en caso de que así lo indiquen.

    Las reglas en cuestión están definidas dentro del código.

In [11]:
def pre_processing(vocab): 
  ### Aquí inicia tu código ###
  ## Lista de stopwords
  stopwords = ["a", "about", "above", "after", "again", "against", "all", "am", 
               "an", "and", "any", "are", "aren't", "as", "at", "be", 
               "because", "been", "before", "being", "below", "between", 
               "both", "but", "by", "can't", "cannot", "could", "couldn't", 
               "did", "didn't", "do", "does", "doesn't", "doing", "don't", 
               "down", "during", "each", "few", "for", "from", "further", 
               "had", "hadn't", "has", "hasn't", "have", "haven't", "having", 
               "he", "he'd", "he'll", "he's", "her", "here", "here's", "hers", 
               "herself", "him", "himself", "his", "how", "how's", "i", "i'd", 
               "i'll", "i'm", "i've", "if", "in", "into", "is", "isn't", "it", 
               "it's", "its", "itself", "let's", "me", "more", "most", 
               "mustn't", "my", "myself", "no", "nor", "not", "of", "off", 
               "on", "once", "only", "or", "other", "ought", "our", "ours", 
               "ourselves", "out", "over", "own", "same", "shan't", "she", 
               "she'd", "she'll", "she's", "should", "shouldn't", "so", "some", 
               "such", "than", "that", "that's", "the", "their", "theirs", 
               "them", "themselves", "then", "there", "there's", "these", 
               "they", "they'd", "they'll", "they're", "they've", "this", 
               "those", "through", "to", "too", "under", "until", "up", "very", 
               "was", "wasn't", "we", "we'd", "we'll", "we're", "we've", 
               "were", "weren't", "what", "what's", "when", "when's", "where", 
               "where's", "which", "while", "who", "who's", "whom", "why", 
               "why's", "with", "won't", "would", "wouldn't", "you", "you'd", 
               "you'll", "you're", "you've", "your", "yours", "yourself", 
               "yourselves"]

  ## Se filtran las stopwords
  for word in vocab:
    ## se pasa a minuscula para comparar con la lista
    if word.lower() in stopwords:
      ## dado que cada palabra está una sola vez en vocab,
      ## remove() es suficiente.
      vocab.remove(word)  

  output = []

  ## Ahora se procesa la parte de las conjugaciones y sufijos, cortando los 
  ## finales de los strings para tratar de dejar solamente las raices.
  for word in vocab:
    if (word[-4:].lower() == "sses"):
      word = word[0:-2]

    if (word[-3:].lower() == "ies"):
      word = word[0:-2]

    if (word[-2:].lower() == "ss"):
      word = word

    if (word[-1:].lower() == "s"):
      word = word[0:-1]

    if (word[-3:].lower() == "eed"):
      word = word[0:-1]

    if (word[-2:].lower() == "ed"):
      word = word[0:-2]

    if (word[-3:].lower() == "ing"):
      word = word[0:-3]
      

    if (word[-1:].lower() == "y"):
      word = word[0:-1] + "i"
      
    if (word[-7:].lower() == "ational"):
      word = word[0:-7] + "ate"
      
    if (word[-6:].lower() == "tional"):
      word = word[0:-2]
      
    if (word[-4:].lower() == "enci"):
      word = word[0:-1] + "e"
      
    if (word[-4:].lower() == "anci"):
      word = word[0:-1] + "e"
      
    if (word[-4:].lower() == "izer"):
      word = word[0:-1]
      
    if (word[-4:].lower() == "abli"):
      word = word[0:-1] + "e"
      
    if (word[-4:].lower() == "alli"):
      word = word[0:-2]
      
    if (word[-4:].lower() == "entli"):
      word = word[0:-2]
      
    if (word[-3:].lower() == "eli"):
      word = word[0:-2]
      
    if (word[-5:].lower() == "ousli"):
      word = word[0:-2]
      
    if (word[-7:].lower() == "ization"):
      word = word[0:-5] + "e"
      
    if (word[-5:].lower() == "ation"):
      word = word[0:-3] + "e"
      
    if (word[-4:].lower() == "ator"):
      word = word[0:-2] + "e"
      
    if (word[-5:].lower() == "alism"):
      word = word[0:-3]
      
    if (word[-7:].lower() == "iveness"):
      word = word[0:-4]
      
    if (word[-7:].lower() == "fulness"):
      word = word[0:-4]
      
    if (word[-7:].lower() == "ousness"):
      word = word[0:-4]
      
    if (word[-5:].lower() == "aliti"):
      word = word[0:-3]
      
    if (word[-5:].lower() == "iviti"):
      word = word[0:-3] + "e"
      
    if (word[-6:].lower() == "biliti"):
      word = word[0:-5] + "le"
      

    if (word[-5:].lower() == "icate"):
      word = word[0:-3]
      
    if (word[-5:].lower() == "ative"):
      word = word[0:-5]
      
    if (word[-5:].lower() == "alize"):
      word = word[0:-3]
      
    if (word[-5:].lower() == "iciti"):
      word = word[0:-3]
      
    if (word[-4:].lower() == "ical"):
      word = word[0:-2]
      
    if (word[-3:].lower() == "ful"):
      word = word[0:-3]
      
    if (word[-4:].lower() == "ness"):
      word = word[0:-4]
      

    if (word[-2:].lower() == "al"):
      word = word[0:-2]
      
    if (word[-4:].lower() == "ance"):
      word = word[0:-4]
      
    if (word[-4:].lower() == "ence"):
      word = word[0:-4]
      
    if (word[-2:].lower() == "er"):
      word = word[0:-2]
      
    if (word[-2:].lower() == "ic"):
      word = word[0:-2]
      
    if (word[-4:].lower() == "able"):
      word = word[0:-4]
      
    if (word[-4:].lower() == "ible"):
      word = word[0:-4]
      
    if (word[-3:].lower() == "ant"):
      word = word[0:-3]
      
    if (word[-5:].lower() == "ement"):
      word = word[0:-5]
      
    if (word[-4:].lower() == "ment"):
      word = word[0:-4]
      
    if (word[-3:].lower() == "ent"):
      word = word[0:-3]
      
    if (word[-3:].lower() == "ion"):
      word = word[0:-3]
      
    if (word[-2:].lower() == "ou"):
      word = word[0:-2]
      
    if (word[-3:].lower() == "ism"):
      word = word[0:-3]
      
    if (word[-3:].lower() == "ate"):
      word = word[0:-3]
      
    if (word[-3:].lower() == "iti"):
      word = word[0:-3]
      
    if (word[-3:].lower() == "ous"):
      word = word[0:-3]
      
    if (word[-3:].lower() == "ive"):
      word = word[0:-3]
      
    if (word[-3:].lower() == "ize"):
      word = word[0:-3]
      

    if (word[-1:].lower() == "e"):
      word = word[0:-1]

    output.append(word)
      
  return output
  ### Aquí termina tu código ###

In [12]:
vocab = pre_processing(vocab)
vocab

['lik', 'human', 'languag', 'programm', 'Spanish', 'mi', 'favorit', 'languag']

` `  
` ` 

**Ejercicio 3 - *Bag of Words* 🐶🐈(0.5 puntos).** 
` `  
` ` 

Considere el siguiente corpus, donde cada elemento del arreglo representa un documento:

In [13]:
d0 = 'El perro se come la comida y después se duerme'
d1 = 'El perro se despierta y después empieza a ladrar'
d2 = 'El perro ladra y después se come la comida'
d3 = 'El gato se come la comida y después se duerme'
d4 = 'El gato se despierta y después empieza a maullar'
d5 = 'El gato maulla y después se come la comida'
dataset = [d0, d1, d2, d3, d4, d5]

El objetivo de este ejercicio es determinar cuáles de  los documentos entregados son los más similares entre sí. Para ello utilizaremos la técnica TF-IDF. 

Como los algoritmos de Machine Learning no comprenden el texto en lenguaje natural, estos documentos deben ser convertidos a vectores numéricos. La representación más simple vista en clases es el **Bag of Words**, método mediante el cuál se cuentan las apariciones de cada palabra en cada uno de los documentos entregados.

Implemente la función **`bag_of_words()`**, que reciba como input un arreglo de documentos y devuelva un pandas dataframe con la representación Bag of Words de los documentos entregados. En esta representación las columnas son el vocabulario y las filas representa las apariciones de cada una de las palabras en los documentos. En otras palabras, cada fila representa el bow de un determinado documento.


Por ejemplo para el siguiente dataset: 

```
dataset = ['El perro ladra', 'El perro come']
```

Debiese entregarnos lo siguiente:


|   | el | perro | ladra | come |
|---|----|-------|------|-------|
| 0 | 1  | 1     | 1    | 0     |
| 1 | 1  | 1     | 0    | 1     |



In [14]:
def bag_of_words(dataset):
  ### Aquí inicia tu código ###
  # Arreglo con los conteo de cada documento
  data = []
  types = []
  doc_tokens_list =[]

  # Se colocan todos los tipos en "type"
  for document in dataset:
    doc_tokens = get_tokens(document)
    doc_tokens_list.append(doc_tokens)
    for token in doc_tokens:
      sub_data = []
      if token not in types:
        types.append(token)

  # Ya encontrados los tipos se va viendo cuantas veces se 
  # encuentran en los documentos
  for doc_tokens in doc_tokens_list:
    sub_data = [0] * len(types)
    for token in doc_tokens:
      sub_data[types.index(token)] += 1
    data.append(sub_data)

  return pd.DataFrame(data, columns=types, 
                      index=["d0", "d1", "d2", "d3", "d4", "d5"])
  ### Aquí termina tu código ###

***Test:***

In [15]:
dataset_bow = bag_of_words(dataset)
dataset_bow

El  perro  se  come  la  comida  ...  a  ladrar  ladra  gato  maullar  maulla
d0   1      1   2     1   1       1  ...  0       0      0     0        0       0
d1   1      1   1     0   0       0  ...  1       1      0     0        0       0
d2   1      1   1     1   1       1  ...  0       0      1     0        0       0
d3   1      0   2     1   1       1  ...  0       0      0     1        0       0
d4   1      0   1     0   0       0  ...  1       0      0     1        1       0
d5   1      0   1     1   1       1  ...  0       0      0     1        0       1

[6 rows x 17 columns]

``
``

***Resultado esperado***: 

|    | El | perro | se | come | la | comida | y | después | duerme | despierta | empieza | a | ladrar | ladra | gato | maullar | maulla |
|----|---:|------:|---:|-----:|---:|-------:|--:|--------:|-------:|----------:|--------:|--:|-------:|------:|-----:|--------:|-------:|
| d0 |  1 |     1 |  2 |    1 |  1 |      1 | 1 |       1 |      1 |         0 |       0 | 0 |      0 |     0 |    0 |       0 |      0 |
| d1 |  1 |     1 |  1 |    0 |  0 |      0 | 1 |       1 |      0 |         1 |       1 | 1 |      1 |     0 |    0 |       0 |      0 |
| d2 |  1 |     1 |  1 |    1 |  1 |      1 | 1 |       1 |      0 |         0 |       0 | 0 |      0 |     1 |    0 |       0 |      0 |
| d3 |  1 |     0 |  2 |    1 |  1 |      1 | 1 |       1 |      1 |         0 |       0 | 0 |      0 |     0 |    1 |       0 |      0 |
| d4 |  1 |     0 |  1 |    0 |  0 |      0 | 1 |       1 |      0 |         1 |       1 | 1 |      0 |     0 |    1 |       1 |      0 |
| d5 |  1 |     0 |  1 |    1 |  1 |      1 | 1 |       1 |      0 |         0 |       0 | 0 |      0 |     0 |    1 |       0 |      1 |

``
``

` `  
` ` 

**Ejercicio 4 - *Calcular TF* (0.5 puntos):** Ahora debemos usar el dataframe del ejercicio anterior para calcular la matriz de TF normalizada por la máxima frecuencia ${max_i({\text{tf}_{i,j})}}$, donde
i corresponde al índice de las filas (bow) y j al de las columnas (palabras). Es decir, dividir cada bow en la cantidad de veces de la palabra que aparezca más veces en ese vector. 


$$\text{nft}_{i,j} = \frac{\text{tf}_{i,j}}{max_i({\text{tf}_{i,j})}}$$

In [16]:
max_i = dataset_bow.max(axis=1)[0]
max_i

2

In [17]:
def calc_tf(dataset_bow):
    df = dataset_bow
    ### Aquí inicia tu código ###
    # Se calcula el maximo de cada fila
    max_i = df.max(axis=1)
    # Se divide cada valor por el maximo de su fila
    for i in range(df.shape[0]):
      df.iloc[i] = df.iloc[i].apply(lambda x: x/max_i[i])
    return df
    ### Aquí termina tu código ###

***Test:***

In [18]:
tf = calc_tf(dataset_bow)
tf

El  perro   se  come   la  ...  ladrar  ladra  gato  maullar  maulla
d0  0.5    0.5  1.0   0.5  0.5  ...     0.0    0.0   0.0      0.0     0.0
d1  1.0    1.0  1.0   0.0  0.0  ...     1.0    0.0   0.0      0.0     0.0
d2  1.0    1.0  1.0   1.0  1.0  ...     0.0    1.0   0.0      0.0     0.0
d3  0.5    0.0  1.0   0.5  0.5  ...     0.0    0.0   0.5      0.0     0.0
d4  1.0    0.0  1.0   0.0  0.0  ...     0.0    0.0   1.0      1.0     0.0
d5  1.0    0.0  1.0   1.0  1.0  ...     0.0    0.0   1.0      0.0     1.0

[6 rows x 17 columns]

In [19]:
list(tf.columns)

['El',
 'perro',
 'se',
 'come',
 'la',
 'comida',
 'y',
 'después',
 'duerme',
 'despierta',
 'empieza',
 'a',
 'ladrar',
 'ladra',
 'gato',
 'maullar',
 'maulla']

``
``

***Resultado esperado***: 

|    |  El | perro |  se | come |  la | comida |   y | después | duerme | despierta | empieza |   a | ladrar | ladra | gato | maullar | maulla |
|----|----:|------:|----:|-----:|----:|-------:|----:|--------:|-------:|----------:|--------:|----:|-------:|------:|-----:|--------:|-------:|
| d0 | 0.5 |   0.5 | 1.0 |  0.5 | 0.5 |    0.5 | 0.5 |     0.5 |    0.5 |       0.0 |     0.0 | 0.0 |    0.0 |   0.0 |  0.0 |     0.0 |    0.0 |
| d1 | 1.0 |   1.0 | 1.0 |  0.0 | 0.0 |    0.0 | 1.0 |     1.0 |    0.0 |       1.0 |     1.0 | 1.0 |    1.0 |   0.0 |  0.0 |     0.0 |    0.0 |
| d2 | 1.0 |   1.0 | 1.0 |  1.0 | 1.0 |    1.0 | 1.0 |     1.0 |    0.0 |       0.0 |     0.0 | 0.0 |    0.0 |   1.0 |  0.0 |     0.0 |    0.0 |
| d3 | 0.5 |   0.0 | 1.0 |  0.5 | 0.5 |    0.5 | 0.5 |     0.5 |    0.5 |       0.0 |     0.0 | 0.0 |    0.0 |   0.0 |  0.5 |     0.0 |    0.0 |
| d4 | 1.0 |   0.0 | 1.0 |  0.0 | 0.0 |    0.0 | 1.0 |     1.0 |    0.0 |       1.0 |     1.0 | 1.0 |    0.0 |   0.0 |  1.0 |     1.0 |    0.0 |
| d5 | 1.0 |   0.0 | 1.0 |  1.0 | 1.0 |    1.0 | 1.0 |     1.0 |    0.0 |       0.0 |     0.0 | 0.0 |    0.0 |   0.0 |  1.0 |     0.0 |    1.0 |

``
``

` `  
` ` 

**Ejercicio 5 - *Calcular IDF* (0.5 punto):**


Implementar `calc_idf(dataset_bow)`. Este debe retornar un diccionario en donde las llaves sean las palabras y los valores sean el calculo de cada idf por palabra.

Recordar que $idf_{t_i} = log_{10}\frac{N}{n_i}$ con $N = $ número de documentos y $n_i = $ Número de documentos que contienen la palabra $t_i$ 

In [20]:
def calc_idf(dataset_bow):
    ### Aquí inicia tu código ###

    output = {}

    ## 6 documentos originalmente
    N = 6
    
    ## Recordar que dataset_bow es una matriz que indica
    ## la cantidad de veces que cada palabra se repite en
    ## cada documento. Cada columna tiene la palabra y las
    ## filas representan la cantidad de veces que la palabra
    ## aparece en cada documento.

    ## Se itera sobre las columnas
    for col_name in dataset_bow.columns[0:]:

      ## Se crea un subset con las filas de valores > 0
      docs = dataset_bow.loc[dataset_bow[col_name] > 0, col_name]
      
      ## Se revisa el tamaño del objeto
      n_i = docs.shape[0]
      frac = float(N/n_i)

      idf_ti = np.log10(frac)

      output[col_name] = idf_ti

    return output
    ### Aquí termina tu código ###

***Test:***

In [21]:
idf = calc_idf(dataset_bow)
idf

{'El': 0.0,
 'a': 0.47712125471966244,
 'come': 0.17609125905568124,
 'comida': 0.17609125905568124,
 'despierta': 0.47712125471966244,
 'después': 0.0,
 'duerme': 0.47712125471966244,
 'empieza': 0.47712125471966244,
 'gato': 0.3010299956639812,
 'la': 0.17609125905568124,
 'ladra': 0.7781512503836436,
 'ladrar': 0.7781512503836436,
 'maulla': 0.7781512503836436,
 'maullar': 0.7781512503836436,
 'perro': 0.3010299956639812,
 'se': 0.0,
 'y': 0.0}

``
``

***Resultado esperado***: 
<table>
    <tr> 
        <td> {'El': 0.0,
 'a': 0.47712125471966244,
 'come': 0.17609125905568124,
 'comida': 0.17609125905568124,
 'despierta': 0.47712125471966244,
 'después': 0.0,
 'duerme': 0.47712125471966244,
 'empieza': 0.47712125471966244,
 'gato': 0.3010299956639812,
 'la': 0.17609125905568124,
 'ladra': 0.7781512503836436,
 'ladrar': 0.7781512503836436,
 'maulla': 0.7781512503836436,
 'maullar': 0.7781512503836436,
 'perro': 0.3010299956639812,
 'se': 0.0,
 'y': 0.0} </td> 
    </tr>
</table> 

``
``

Puede notar el bajo puntaje otorgado a las palabras que más se repiten! 😮

**Ejercicio 6 - *Calcular TF-IDF & concluir similitud de documentos.* (1 punto)**


Programe la función `calc_tf_idf(tf, idf)` que entrega el dataframe TF-IDF asociado al dataset que estamos analizando.

In [22]:
def calc_tf_idf(tf, idf):
    ### Aquí inicia tu código ###
    # Hacemos una copia para modificar
    df = tf
    columns = list(df.columns)
    # A cada columna se aplica la funcion TF-IDF
    for column in columns:
      df[column] = df[column].apply(lambda x: x*idf.get(column))
    return df
    ### Aquí termina tu código ### 

***Test.***

In [23]:
tf_idf = calc_tf_idf(tf, idf)
tf_idf

El     perro   se      come  ...     ladra      gato   maullar    maulla
d0  0.0  0.150515  0.0  0.088046  ...  0.000000  0.000000  0.000000  0.000000
d1  0.0  0.301030  0.0  0.000000  ...  0.000000  0.000000  0.000000  0.000000
d2  0.0  0.301030  0.0  0.176091  ...  0.778151  0.000000  0.000000  0.000000
d3  0.0  0.000000  0.0  0.088046  ...  0.000000  0.150515  0.000000  0.000000
d4  0.0  0.000000  0.0  0.000000  ...  0.000000  0.301030  0.778151  0.000000
d5  0.0  0.000000  0.0  0.176091  ...  0.000000  0.301030  0.000000  0.778151

[6 rows x 17 columns]

``
``

***Resultado esperado***: 

|    |  El |    perro |  se |     come |       la |   comida |   y | después |   duerme | despierta |  empieza |        a |   ladrar |    ladra |     gato |  maullar |   maulla |
|----|----:|---------:|----:|---------:|---------:|---------:|----:|--------:|---------:|----------:|---------:|---------:|---------:|---------:|---------:|---------:|---------:|
| d0 | 0.0 | 0.150515 | 0.0 | 0.088046 | 0.088046 | 0.088046 | 0.0 |     0.0 | 0.238561 |  0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 |
| d1 | 0.0 | 0.301030 | 0.0 | 0.000000 | 0.000000 | 0.000000 | 0.0 |     0.0 | 0.000000 |  0.477121 | 0.477121 | 0.477121 | 0.778151 | 0.000000 | 0.000000 | 0.000000 | 0.000000 |
| d2 | 0.0 | 0.301030 | 0.0 | 0.176091 | 0.176091 | 0.176091 | 0.0 |     0.0 | 0.000000 |  0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.778151 | 0.000000 | 0.000000 | 0.000000 |
| d3 | 0.0 | 0.000000 | 0.0 | 0.088046 | 0.088046 | 0.088046 | 0.0 |     0.0 | 0.238561 |  0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.150515 | 0.000000 | 0.000000 |
| d4 | 0.0 | 0.000000 | 0.0 | 0.000000 | 0.000000 | 0.000000 | 0.0 |     0.0 | 0.000000 |  0.477121 | 0.477121 | 0.477121 | 0.000000 | 0.000000 | 0.301030 | 0.778151 | 0.000000 |
| d5 | 0.0 | 0.000000 | 0.0 | 0.176091 | 0.176091 | 0.176091 | 0.0 |     0.0 | 0.000000 |  0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.301030 | 0.000000 | 0.778151 |


``
``

Ahora que tenemos el dataframe de TF-IDF, nos queda calcular la similitud coseno entre todos los vectores. Notar que la matriz resultante será una matriz simétrica. Implemente la función *cosine_similarity(v1, v2)* que recibe dos vectores (v1 y v2) y calcula la similitud coseno entre ambos vectores. Concluya cuáles son los dos documentos más similares.

In [24]:
def cosine_similarity(v1, v2):
  ### Aquí inicia tu código ###
  # Se calcula el producto punto
  a = np.dot(v1,v2)
  # Se calcula el modulo de cada vector y se multiplica
  b = np.linalg.norm(v1) * np.linalg.norm(v2)
  # Finalemente se divide ambos resultados anteriores
  return a/b
  ### Aquí termina tu código ### 

similarity_matrix = np.zeros((6,6))
for i, v1 in enumerate(tf_idf.index.values):
  for j, v2 in enumerate(tf_idf.index.values):
      similarity = cosine_similarity(tf_idf.loc[v1].values, tf_idf.loc[v2].values)
      similarity_matrix[i][j] = similarity
print(similarity_matrix)

[[1.         0.12032418 0.3223436  0.77967014 0.         0.1632828 ]
 [0.12032418 1.         0.08686457 0.         0.4952126  0.        ]
 [0.3223436  0.08686457 1.         0.1632828  0.         0.11787732]
 [0.77967014 0.         0.1632828  1.         0.12032418 0.3223436 ]
 [0.         0.4952126  0.         0.12032418 1.         0.08686457]
 [0.1632828  0.         0.11787732 0.3223436  0.08686457 1.        ]]


    Escriba su respuesta aquí

    El documento d0 y el documento d3 son los mas similares con una simulitud de
    0.78 aproximadamente.

![gato](https://www.elagoradiario.com/wp-content/uploads/2020/05/Gato-mascarilla.jpg)

**Cualquier recomendación que nos quisieran dar para una futura tarea es bienvenid@!**